packages/zowe-explorer/src/dataset/actions.ts

```diff
@@ -144,20 +145,36 @@ export async function uploadDialog(node: ZoweDatasetNode, datasetProvider: api.I
         openLabel: "Upload File",
         canSelectMany: true,
     };
-
     const value = await api.Gui.showOpenDialog(fileOpenOptions);
-
-    if (value && value.length) {
-        await Promise.all(
-            value.map(async (item) => {
-                // Convert to vscode.TextDocument
-                const doc = await vscode.workspace.openTextDocument(item);
-                await uploadFile(node, doc);
-            })
+    if (value?.length > 0) {
+        await api.Gui.withProgress(
+            {
+                location: vscode.ProgressLocation.Notification,
+                title: localize("uploadFile.response.upload.title", "Uploading to data set"),
+                cancellable: true,
+            },
+            async (progress, token) => {
+                let index = 0;
+                for (const item of value) {
+                    if (token.isCancellationRequested) {
+                        api.Gui.showMessage(localize("uploadFile.uploadCancelled", "Upload action was cancelled."));
+                        break;
+                    }
+                    reportProgress(progress, value.length, index, "Uploading");
+                    const doc = await vscode.workspace.openTextDocument(item);
```

Comment:
For putContents all we need is the file path. Calling the VS Code openTextDocument() has a big performance impact. I can see in my log files that loading it as a document will actually call the language server, if you have one installed for the language of the file being uploaded, sending the file around, fully parsing it, etc. I would just hand down the file path to upload and nothing else as the files will not be opened in an editor at all.

Reply:
addressed not opening doc to get path when item itself was holding the information in commit https://github.com/zowe/vscode-extension-for-zowe/pull/2185/commits/11195f8623ebbe500ae2ac1e89b911cb5c4d5f7a

with the zOSMF specific issue, should we create a bug with zOSMF rest api?

Reply:
I've reached out to the zOSMF team in regards to the zosmf type specific bug.